libraries/bot-applicationinsights/src/test/java/com/microsoft/bot/applicationinsights/TelemetryWaterfallTests.java

```diff
@@ -0,0 +1,396 @@
+        });
+
+        MyWaterfallDialog waterfallDialog = new MyWaterfallDialog("test", steps);
+
+        dialogs.add(waterfallDialog);
+        dialogs.setTelemetryClient(telemetryClient);
+
+        new TestFlow(adapter, turnContext -> {
+            DialogContext dc = dialogs.createContext(turnContext).join();
+            dc.continueDialog().join();
+            if (!turnContext.getResponded()) {
+                dc.beginDialog("test", null).join();
+            }
+            return CompletableFuture.completedFuture(null);
+        })
+            .send("hello")
+            .assertReply("step1")
+            .send("hello")
+            .assertReply("step2")
+            .send("hello")
+            .assertReply("step1")
+            .startTest()
+            .join();
+
+        Assert.assertEquals((Integer)7, counter[0]);
+
+        // Verify:
+        // Event name is "WaterfallCancel"
+        // Event occurs on the 4th event logged
+        // Event contains DialogId
+        // Event DialogId is set correctly.
+        Assert.assertTrue(saved_properties.get("WaterfallStart_0").containsKey("DialogId"));
+        Assert.assertTrue(saved_properties.get("WaterfallStart_0").containsKey("InstanceId"));
+        Assert.assertTrue(saved_properties.get("WaterfallCancel_4").containsKey("DialogId"));
+        Assert.assertEquals("test", saved_properties.get("WaterfallCancel_4").get("DialogId"));
+        Assert.assertTrue(saved_properties.get("WaterfallCancel_4").containsKey("StepName"));
+        Assert.assertTrue(saved_properties.get("WaterfallCancel_4").containsKey("InstanceId"));
+
+        // Event contains "StepName"
+        // Event naming on lambda's is "StepXofY"
+        Assert.assertEquals("Step3of3", saved_properties.get("WaterfallCancel_4").get("StepName"));
+        Assert.assertTrue(waterfallDialog.getCancelDialogCalled());
+        Assert.assertFalse(waterfallDialog.getEndDialogCalled());
+    }
+
+    private static List<WaterfallStep> newWaterfall() {
+        List<WaterfallStep> waterfall = new ArrayList<>();
+
+        waterfall.add(step -> {
+            step.getContext().sendActivity("step1");
+            return CompletableFuture.completedFuture(Dialog.END_OF_TURN);
+        });
+
+        waterfall.add(step -> {
+            step.getContext().sendActivity("step2");
+            return CompletableFuture.completedFuture(Dialog.END_OF_TURN);
+        });
+
+        waterfall.add(step -> {
+            step.getContext().sendActivity("step3");
+            return CompletableFuture.completedFuture(Dialog.END_OF_TURN);
+        });
+
+        return waterfall;
+    }
+
+    private class MyWaterfallDialog extends WaterfallDialog {
+        private Boolean endDialogCalled;
+        private Boolean cancelDialogCalled;
+
+        public MyWaterfallDialog(String id, List<WaterfallStep> actions) {
+            super(id, actions);
+
+            endDialogCalled = false;
+            cancelDialogCalled = false;
```

Comment:
Can we move this to the properties? Just like [C#](https://github.com/microsoft/botbuilder-dotnet/blob/main/tests/Microsoft.Bot.Builder.ApplicationInsights.Tests/TelemetryWaterfallTests.cs#L266)